File: bot-ui/src/overlay/Overlay.ts
import { OverlayElement } from "@/overlay/OverlayElement";
import Resolution from "@/overlay/Resolution";

export default class Overlay {
  id: string;
  title: string;
  resolution: Resolution;
  elements: OverlayElement[];

  constructor(
    id: string,
    title: string,
    resolution: Resolution,
    elements: OverlayElement[] = []
  ) {
    this.id = id;
    this.title = title;
    this.resolution = resolution;
    this.elements = elements;
  }
}

File: bot-core/src/main/kotlin/fr/delphes/bot/overlay/Overlay.kt
package fr.delphes.bot.overlay

import kotlinx.serialization.Serializable

@Serializable
data class Overlay(
    val id: String,
    val title: String,
    val resolution: Resolution,
    val elements: List<OverlayElement<*>> = emptyList(),
) {
    fun sanitize(): Overlay {
        return this.copy(
            elements = elements.reorder()
        )
    }

    private fun List<OverlayElement<*>>.reorder() = this
        .sortedBy { it.general.sortOrder }
        .mapIndexed { index, element -> element.copy(general = element.general.copy(sortOrder = index)) }
}
